lib/semantics/resolve-labels.cc

```diff
@@ -898,12 +898,12 @@ void CheckLabelDoConstraints(const SourceStmtList &dos,
               TargetStatementEnum::CompatibleDo)) {
         errorHandler.Say(doTarget.parserCharBlock,
             parser::MessageFormattedText{
-                "'%u' DO statements must terminate with END DO or CONTINUE"_err_en_US,
+                "'%u' Obsolete construct.  A DO loop should terminate with END DO or CONTINUE"_err_en_US,
```

Comment:
The construct is obsolete in either case.  Maybe you could use the same text in both branches.

How about "Only an END DO or CONTINUE statement should be used to terminate a labeled DO loop."

Reply:
That seems a reasonable message.  

There's little utility to putting in the message something to the effect that "an assignment (or ...) statement would also be fine, but only if you're not trying to have strict 2018 standard conformance".

Reply:
After talking to @schweitzpgi, I realized that the error message might come happen for real errors rather than just obsolete constructs, so we don't want to include that text.

It seems weird to me to have the same text for both the error and warning cases, but I can't think of a better alternative.

Reply:
You could say `Only an END DO or CONTINUE should be used...` for the warning
and `Only an END DO or CONTINUE may be used...` for the error.
"Should" implies it is a recommendation, not a requirement.

Reply:
A possible alternative to
  Only <...> may be used
is
  <...> must be used
(I'm currently using the "must" for in new messages.)

Reply:
Thanks, @tskeith and @vdonaldson.  I'll update the messages as per @vdonaldson's recommendation.